src/poly/multiopen/verifier.rs

```diff
@@ -0,0 +1,246 @@
+use super::super::{
+    commitment::{Guard, Params, MSM},
+    Error,
+};
+use super::{Proof, VerifierQuery};
+use crate::arithmetic::{
+    eval_polynomial, get_challenge_scalar, lagrange_interpolate, Challenge, CurveAffine, Field,
+};
+use crate::plonk::hash_point;
+use crate::transcript::Hasher;
+use std::collections::{BTreeMap, BTreeSet};
+
+#[derive(Debug, Clone)]
+struct CommitmentData<C: CurveAffine> {
+    set_index: usize,
+    point_indices: Vec<usize>,
+    evals: Vec<C::Scalar>,
+}
+
+impl<'a, C: CurveAffine> Proof<C> {
+    /// Verify a multi-opening proof
+    pub fn verify<I, HBase: Hasher<C::Base>, HScalar: Hasher<C::Scalar>>(
+        &self,
+        params: &'a Params<C>,
+        transcript: &mut HBase,
+        transcript_scalar: &mut HScalar,
+        queries: I,
+        msm: MSM<'a, C>,
+    ) -> Result<Guard<'a, C>, Error>
+    where
+        I: IntoIterator<Item = VerifierQuery<'a, C>> + Clone,
+    {
+        // Sample x_4 for compressing openings at the same points together
```

Comment:
The beginning of this function should have

```rust
        // Scale the MSM by a random factor to ensure that if the existing MSM
        // has is_zero() == false then this argument won't be able to interfere
        // with it to make it true, with high probability.
        msm.scale(C::Scalar::random());
```

and the existing code in `plonk/verifier.rs` should be removed because it is now no longer the PLONK implementation's responsibility to ensure these are linearly independent, and in any event usage of the multiopen API on its own would be unsafe.